### DISCUSSION TITLE
Empty dependency files make Trivy fail

### DISCUSSION THREAD

**Discussion by tcoupin** [aquasecurity/trivy#4295] [Q&A]
## Description

Seems that #472 is back.

When trivy comes across a application dependency file which is empty, it will fail (in this case `composer.lock`).

## What did you expect to happen?

For Trivy to ignore the file and maybe issue a warning.

## What happened instead?

Trivy failed with a FATAL error.

## Output of run with `-debug`:

```
2023-04-06T15:42:21.857Z        FATAL   image scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:427
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:266
  - scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:679
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        /home/runner/work/trivy/trivy/pkg/scanner/scan.go:146
  - analyze error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/image.Artifact.Inspect
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/image/image.go:139
  - failed to analyze layer (sha256:f706995fe867983072efeee3e5761f3a48eff6d57dd50b80535edebe2b37268c):
    github.com/aquasecurity/trivy/pkg/fanal/artifact/image.Artifact.inspect.func1
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/image/image.go:242
  - post analysis error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/image.Artifact.inspectLayer
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/image/image.go:323
  - post analysis error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer.AnalyzerGroup.PostAnalyze
        /home/runner/work/trivy/trivy/pkg/fanal/analyzer/analyzer.go:487
  - composer walk error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/php/composer.composerAnalyzer.PostAnalyze
        /home/runner/work/trivy/trivy/pkg/fanal/analyzer/language/php/composer/composer.go:73
  - parse error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer/language/php/composer.composerAnalyzer.PostAnalyze.func2
        /home/runner/work/trivy/trivy/pkg/fanal/analyzer/language/php/composer/composer.go:58
  - failed to parse opt/drupal/web/core/modules/system/tests/fixtures/HtaccessTest/composer.lock:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer/language.Parse
        /home/runner/work/trivy/trivy/pkg/fanal/analyzer/language/analyze.go:50
  - decode error:
    github.com/aquasecurity/go-dep-parser/pkg/php/composer.(*Parser).Parse
        /home/runner/go/pkg/mod/github.com/aquasecurity/go-dep-parser@v0.0.0-20230324043952-2172dc218241/pkg/php/composer/parse.go:43
  - EOF
```

## Output of `trivy -v`:

```
Version: 0.39.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-04-06 12:08:08.048185529 +0000 UTC
  NextUpdate: 2023-04-06 18:08:08.048184829 +0000 UTC
  DownloadedAt: 2023-04-06 15:29:09.625988232 +0000 UTC
```

## Additional details (base image name, container registry info...):

image `drupal:9.4.1-apache-bullseye`

**Comment by DmitriyLewen** [comment]
Hello @tcoupin 
Thanks for your report!

Looks like problem is not in Trivy.
`composer.lock` is json file. Empty file is invalid for `json`.
`json` file should contain `{}`.

___

Drupal added `{}` to files in 10.1 (https://www.drupal.org/project/drupal/issues/3138691#comment-14928103).
I checked. Actual dev version contains corrected files:
```bash
➜  cat ./drupal-10.1.x-dev/core/modules/system/tests/fixtures/HtaccessTest/composer.json
{}
```
___

Before release `10.1` you can [skip](https://aquasecurity.github.io/trivy/v0.39/docs/vulnerability/examples/others/#skip-files) this file.
I also saw that this file placed in `test` folder. I am not sure, but perhaps this folder contains only test files, that eed to be skipped during scan.

Regards, Dmitriy

**Comment by tcoupin** [comment]
Thanks you Dimitry.

I don't understand too if this composer.lock is useful.
But in my previous CI pipeline, trivy wasn't failed. I can say if this
change of behavior comes from change in the Drupal image or in trivy.

Le ven. 7 avr. 2023 à 05:40, DmitriyLewen ***@***.***> a
écrit :

**Comment by DmitriyLewen** [comment]
More likely you got error but in debug log.

We updated logic for `composer.lock`.
New post analyzer logic return errors for invalid files.
We have similar issue with `jar` files.

We are working on adding flag to ignore analyze errors - https://github.com/aquasecurity/trivy/pull/3988
